File: src/store.js
/**
 * src/store.js — atomic JSON persistence for the bot.
 *
 * Everything (chats, drafts, schedules, channels, templates, settings) lives
 * in ONE JSON file which is rewritten atomically: write to a sibling temp file
 * then `rename()`, so a crash mid-write can never corrupt live state.
 *
 * On a platform with an ephemeral filesystem (Veroa, Render, Railway …) the
 * file survives restarts of the *same* instance but not redeploys — point
 * APB_DATA_DIR at a mounted disk to keep it forever.
 */

import fs from 'node:fs';
import os from 'node:os';
import path from 'node:path';

const REPEAT_SECONDS = { hourly: 3600, daily: 86400, weekly: 7 * 86400 };

let seq = 0;

function newId(prefix = '') {
  seq += 1;
  return `${prefix}${Date.now()}-${seq}`;
}

export function defaultDataDir(projectRoot) {
  return process.env.APB_DATA_DIR || process.env.DATA_DIR || path.join(projectRoot, 'data');
}

/**
 * Make sure `dir` exists and is writable.  Falls back to a temp directory with
 * a loud warning instead of crash-looping on a read-only filesystem.
 */
export function ensureWritableDir(dir, warn = console.warn) {
  const probe = path.join(dir, '.apb-write-test');
  try {
    fs.mkdirSync(dir, { recursive: true });
    fs.writeFileSync(probe, 'ok');
    fs.unlinkSync(probe);
    return dir;
  } catch (err) {
    const fallback = path.join(os.tmpdir(), 'apb-data');
    warn(`WARNING: data dir ${dir} is not writable (${err.message}).\n`
      + `         Falling back to ${fallback} — state is lost when this instance restarts.\n`
      + '         Mount a disk/volume and set APB_DATA_DIR to keep it.');
    fs.mkdirSync(fallback, { recursive: true });
    return fallback;
  }
}

export class Store {
  static DEFAULTS = {
    chats: {},      // chat_id -> {title, type, first_seen}
    drafts: {},     // draft_id -> {title, markdown, mode, buttons, media, created}
    scheduled: {},  // job_id -> {run_at, target, markdown, mode, buttons, media, note, status, repeat}
    channels: {},   // chat_id -> {title, signature, delay, added}
    templates: {},  // tpl_id -> {name, markdown, buttons, media, signature, created}
    settings: { turbo: false, watermark_on: false, watermark_text: '' },
    admins: [],
    sent: 0,
    failed: 0,
  };

  constructor(filePath) {
    this.path = filePath;
    this.data = structuredClone(Store.DEFAULTS);
    this.load();
  }

  // ------------------------------------------------------------------- io

  load() {
    try {
      const parsed = JSON.parse(fs.readFileSync(this.path, 'utf8'));
      if (parsed && typeof parsed === 'object') {
        for (const [key, value] of Object.entries(parsed)) this.data[key] = value;
      }
    } catch (err) {
      if (err.code !== 'ENOENT') {
        // Corrupt JSON: keep a copy for forensics, start fresh, never crash.
        try {
          fs.renameSync(this.path, `${this.path}.corrupt-${Date.now()}`);
        } catch { /* ignore */ }
      }
    }
  }

  save() {
    const dir = path.dirname(path.resolve(this.path));
    fs.mkdirSync(dir, { recursive: true });
    const tmp = path.join(dir, `.apb-${process.pid}-${Date.now()}.tmp`);
    fs.writeFileSync(tmp, JSON.stringify(this.data, null, 1));
    fs.renameSync(tmp, this.path);
  }

  // -------------------------------------------------------------- settings

  setting(key, fallback = null) {
    return this.data.settings?.[key] ?? fallback;
  }

  setSetting(key, value) {
    if (!this.data.settings) this.data.settings = {};
    this.data.settings[key] = value;
    this.save();
  }

  // -------------------------------------------------------------- channels

  addChannel(chatId, { title = '', signature = null, delay = null } = {}) {
    const key = String(chatId);
    const entry = this.data.channels[key] || (this.data.channels[key] = {});
    entry.title = title || entry.title || key;
    entry.added = entry.added || Date.now() / 1000;
    if (signature !== null) entry.signature = signature;
    if (delay !== null) entry.delay = Number(delay);
    this.save();
    return entry;
  }

  channel(chatId) {
    return this.data.channels[String(chatId)] || null;
  }

  channels() {
    return { ...this.data.channels };
  }

  delChannel(chatId) {
    const key = String(chatId);
    if (this.data.channels[key]) {
      delete this.data.channels[key];
      this.save();
      return true;
    }
    return false;
  }

  // ------------------------------------------------------------- templates

  addTemplate(name, markdown, { buttons = null, media = null, signature = '' } = {}) {
    const tid = newId('t');
    this.data.templates[tid] = {
      name,
      markdown,
      buttons: buttons || [],
      media: media || [],
      signature: signature || '',
      created: Date.now() / 1000,
    };
    this.save();
    return tid;
  }

  templates() {
    return Object.entries(this.data.templates).sort((a, b) => (b[1].created || 0) - (a[1].created || 0));
  }

  template(tid) {
    return this.data.templates[String(tid)] || null;
  }

  delTemplate(tid) {
    const key = String(tid);
    if (this.data.templates[key]) {
      delete this.data.templates[key];
      this.save();
      return true;
    }
    return false;
  }

  // ----------------------------------------------------------------- chats

  chatEnsure(chatId, title = '', chatType = '') {
    const key = String(chatId);
    const entry = this.data.chats[key] || (this.data.chats[key] = {});
    const changed = entry.title !== (title || entry.title || '');
    entry.title = title || entry.title || '';
    entry.type = chatType || entry.type || '';
    if (!entry.first_seen) entry.first_seen = Date.now() / 1000;
    if (changed || !entry.first_seen) this.save();
    return entry;
  }

  chats() {
    return Object.keys(this.data.chats);
  }

  // ---------------------------------------------------------------- drafts

  addDraft(title, markdown, { mode = 'rich', buttons = null, media = null } = {}) {
    const did = newId();
    this.data.drafts[did] = {
      title,
      markdown,
      mode,
      buttons: buttons || [],
      media: media || [],
      created: Date.now() / 1000,
    };
    this.save();
    return did;
  }

  drafts() {
    return Object.entries(this.data.drafts).sort((a, b) => (b[1].created || 0) - (a[1].created || 0));
  }

  draft(did) {
    return this.data.drafts[String(did)] || null;
  }

  delDraft(did) {
    const key = String(did);
    if (this.data.drafts[key]) {
      delete this.data.drafts[key];
      this.save();
      return true;
    }
    return false;
  }

  // ------------------------------------------------------------- scheduled

  addScheduled(runAt, target, markdown, {
    mode = 'rich', buttons = null, media = null, note = '',
    repeat = 'none', slideshow = false, stars = null,
  } = {}) {
    const jid = newId();
    this.data.scheduled[jid] = {
      run_at: runAt,
      target,
      markdown,
      mode,
      buttons: buttons || [],
      media: media || [],
      note,
      status: 'pending',
      repeat,
      slideshow: !!slideshow,
      stars,
    };
    this.save();
    return jid;
  }

  repeatInterval(job) {
    const rep = job?.repeat ?? 'none';
    if (REPEAT_SECONDS[rep]) return REPEAT_SECONDS[rep];
    if (typeof rep === 'number' && rep > 0) return rep;
    if (typeof rep === 'string' && rep.startsWith('every:')) {
      const value = parseFloat(rep.split(':')[1]);
      return Number.isFinite(value) ? value : null;
    }
    return null;
  }

  /** Advance a recurring job to its next run (no drift). */
  rescheduleRecurring(jid, now = Date.now() / 1000) {
    const job = this.data.scheduled[String(jid)];
    if (!job) return false;
    const interval = this.repeatInterval(job);
    if (!interval) return false;
    let next = (job.run_at || now) + interval;
    while (next <= now) next += interval;      // catch up if we fell behind
    job.run_at = next;
    job.status = 'pending';
    this.save();
    return true;
  }

  /** Return and mark running all pending jobs whose time has come. */
  dueScheduled(now = Date.now() / 1000) {
    const due = [];
    for (const [jid, job] of Object.entries(this.data.scheduled)) {
      if (job.status === 'pending' && (job.run_at || 0) <= now) {
        job.status = 'running';
        due.push([jid, job]);
      }
    }
    if (due.length) this.save();
    return due;
  }

  finishScheduled(jid, { ok = true, error = null } = {}) {
    const key = String(jid);
    const job = this.data.scheduled[key];
    if (!job) return;
    if (ok) delete this.data.scheduled[key];
    else {
      job.status = 'failed';
      job.error = error || 'unknown error';
    }
    this.save();
  }

  scheduled() {
    return Object.entries(this.data.scheduled).sort((a, b) => (a[1].run_at || 0) - (b[1].run_at || 0));
  }

  delScheduled(jid) {
    const key = String(jid);
    if (this.data.scheduled[key]) {
      delete this.data.scheduled[key];
      this.save();
      return true;
    }
    return false;
  }

  // ----------------------------------------------------------------- stats

  bump(ok = true) {
    if (ok) this.data.sent += 1;
    else this.data.failed += 1;
    this.save();
  }

  stats() {
    return {
      chats: Object.keys(this.data.chats).length,
      drafts: Object.keys(this.data.drafts).length,
      scheduled: Object.keys(this.data.scheduled).length,
      channels: Object.keys(this.data.channels).length,
      templates: Object.keys(this.data.templates).length,
      sent: this.data.sent,
      failed: this.data.failed,
    };
  }
}

export default Store;

File: apb/store.py
"""apb.store — tiny atomic JSON persistence for the bot."""

from __future__ import annotations

import json
import os
import tempfile
import threading
import time


class _Seq:
    """Monotonic suffix so ids never collide inside the same millisecond."""

    def __init__(self):
        self._n = 0
        self._lock = threading.Lock()

    def next(self):
        with self._lock:
            self._n += 1
            return self._n


_SEQ = _Seq()


def _new_id(prefix=""):
    return "{}{}-{}".format(prefix, int(time.time() * 1000), _SEQ.next())


class Store:
    """All state lives in one JSON file, written atomically."""

    DEFAULTS = {
        "chats": {},        # chat_id -> {"title", "type", "first_seen"}
        "drafts": {},       # draft_id -> {"title", "markdown", "mode", "buttons", "media", "created"}
        "scheduled": {},    # job_id  -> {"run_at", "target", "markdown", "mode", "buttons", "media", "note", "status", "repeat"}
        "channels": {},     # chat_id -> {"title", "signature", "delay", "added"}
        "templates": {},    # tpl_id  -> {"name", "markdown", "buttons", "media", "signature", "created"}
        "settings": {"turbo": False, "watermark_on": False, "watermark_text": ""},
        "sent": 0,
        "failed": 0,
    }

    # ------------------------------------------------------- settings

    def setting(self, key, default=None):
        return self.data.get("settings", {}).get(key, default)

    def set_setting(self, key, value):
        with self._lock:
            self.data.setdefault("settings", {})[key] = value
            self.save()

    # ------------------------------------------------------- channels

    def add_channel(self, chat_id, title="", signature=None, delay=None):
        with self._lock:
            entry = self.data["channels"].setdefault(str(chat_id), {})
            entry["title"] = title or entry.get("title", str(chat_id))
            entry.setdefault("added", time.time())
            if signature is not None:
                entry["signature"] = signature
            if delay is not None:
                entry["delay"] = float(delay)
            self.save()
            return entry

    def channel(self, chat_id):
        return self.data["channels"].get(str(chat_id))

    def channels(self):
        return dict(self.data["channels"])

    def del_channel(self, chat_id):
        with self._lock:
            if str(chat_id) in self.data["channels"]:
                del self.data["channels"][str(chat_id)]
                self.save()
                return True
            return False

    # ------------------------------------------------------ templates

    def add_template(self, name, markdown, buttons=None, media=None, signature=""):
        with self._lock:
            tid = _new_id("t")
            self.data["templates"][tid] = {
                "name": name, "markdown": markdown,
                "buttons": buttons or [], "media": media or [],
                "signature": signature or "", "created": time.time(),
            }
            self.save()
            return tid

    def templates(self):
        return sorted(self.data["templates"].items(), key=lambda kv: -kv[1].get("created", 0))

    def template(self, tid):
        return self.data["templates"].get(str(tid))

    def del_template(self, tid):
        with self._lock:
            if str(tid) in self.data["templates"]:
                del self.data["templates"][str(tid)]
                self.save()
                return True
            return False

    def __init__(self, path):
        self.path = path
        self._lock = threading.RLock()
        self.data = json.loads(json.dumps(self.DEFAULTS))  # deep copy
        self._load()

    # ------------------------------------------------------------ io

    def _load(self):
        try:
            with open(self.path, "r", encoding="utf-8") as fh:
                loaded = json.load(fh)
            for key, val in loaded.items():
                self.data[key] = val
        except (OSError, ValueError):
            pass

    def save(self):
        with self._lock:
            directory = os.path.dirname(os.path.abspath(self.path))
            os.makedirs(directory, exist_ok=True)
            fd, tmp = tempfile.mkstemp(dir=directory, suffix=".tmp")
            try:
                with os.fdopen(fd, "w", encoding="utf-8") as fh:
                    json.dump(self.data, fh, ensure_ascii=False, indent=1, sort_keys=True)
                os.replace(tmp, self.path)
            except BaseException:
                try:
                    os.unlink(tmp)
                except OSError:
                    pass
                raise

    # ---------------------------------------------------------- chats

    def chat_ensure(self, chat_id, title="", chat_type=""):
        with self._lock:
            entry = self.data["chats"].setdefault(str(chat_id), {})
            changed = entry.get("title") != (title or entry.get("title", ""))
            entry["title"] = title or entry.get("title", "")
            entry["type"] = chat_type or entry.get("type", "")
            entry.setdefault("first_seen", time.time())
            if changed or "first_seen" not in entry:
                self.save()
            return entry

    def chats(self):
        return list(self.data["chats"].keys())

    # --------------------------------------------------------- drafts

    def add_draft(self, title, markdown, mode="rich", buttons=None, media=None):
        with self._lock:
            did = _new_id()
            self.data["drafts"][did] = {
                "title": title,
                "markdown": markdown,
                "mode": mode,
                "buttons": buttons or [],
                "media": media or [],
                "created": time.time(),
            }
            self.save()
            return did

    def drafts(self):
        return sorted(self.data["drafts"].items(), key=lambda kv: -kv[1].get("created", 0))

    def draft(self, did):
        return self.data["drafts"].get(str(did))

    def del_draft(self, did):
        with self._lock:
            if str(did) in self.data["drafts"]:
                del self.data["drafts"][str(did)]
                self.save()
                return True
            return False

    # ------------------------------------------------------ scheduled

    def add_scheduled(self, run_at, target, markdown, mode="rich", buttons=None,
                      media=None, note="", repeat="none", slideshow=False, stars=None):
        with self._lock:
            jid = _new_id()
            self.data["scheduled"][jid] = {
                "run_at": run_at,
                "target": target,
                "markdown": markdown,
                "mode": mode,
                "buttons": buttons or [],
                "media": media or [],
                "note": note,
                "status": "pending",
                "repeat": repeat,
                "slideshow": bool(slideshow),
                "stars": stars,
            }
            self.save()
            return jid

    REPEAT_SECONDS = {"hourly": 3600, "daily": 86400, "weekly": 7 * 86400}

    def repeat_interval(self, job):
        rep = job.get("repeat") or "none"
        if rep in self.REPEAT_SECONDS:
            return self.REPEAT_SECONDS[rep]
        if isinstance(rep, (int, float)) and rep > 0:
            return rep
        if isinstance(rep, str) and rep.startswith("every:"):
            try:
                return float(rep.split(":", 1)[1])
            except ValueError:
                return None
        return None

    def reschedule_recurring(self, jid, now=None):
        """Advance a recurring job to its next run (no drift)."""
        now = now or time.time()
        with self._lock:
            job = self.data["scheduled"].get(str(jid))
            if job is None:
                return False
            interval = self.repeat_interval(job)
            if not interval:
                return False
            nxt = job.get("run_at", now) + interval
            while nxt <= now:               # catch up if we fell behind
                nxt += interval
            job["run_at"] = nxt
            job["status"] = "pending"
            self.save()
            return True

    def due_scheduled(self, now=None):
        """Return and mark running all pending jobs whose time has come."""
        now = now or time.time()
        due = []
        with self._lock:
            for jid, job in self.data["scheduled"].items():
                if job.get("status") == "pending" and job.get("run_at", 0) <= now:
                    job["status"] = "running"
                    due.append((jid, job))
            if due:
                self.save()
        return due

    def finish_scheduled(self, jid, ok=True, error=None):
        with self._lock:
            job = self.data["scheduled"].get(str(jid))
            if job is None:
                return
            if ok:
                del self.data["scheduled"][str(jid)]
            else:
                job["status"] = "failed"
                job["error"] = error or "unknown error"
            self.save()

    def scheduled(self):
        return sorted(self.data["scheduled"].items(), key=lambda kv: kv[1].get("run_at", 0))

    def del_scheduled(self, jid):
        with self._lock:
            if str(jid) in self.data["scheduled"]:
                del self.data["scheduled"][str(jid)]
                self.save()
                return True
            return False

    # ---------------------------------------------------------- stats

    def bump(self, ok=True):
        with self._lock:
            self.data["sent" if ok else "failed"] += 1
            self.save()

    def stats(self):
        return {
            "chats": len(self.data["chats"]),
            "drafts": len(self.data["drafts"]),
            "scheduled": len(self.data["scheduled"]),
            "sent": self.data["sent"],
            "failed": self.data["failed"],
        }
